x/examples/test-connectivity/main.go

```diff
@@ -121,6 +123,55 @@ func init() {
 	}
 }
 
+func getReportFromTrace(ctx context.Context, r *connectivityReport, hostname string) context.Context {
+	var dnsStart, connectStart time.Time
+	ctx = httptrace.WithClientTrace(ctx, &httptrace.ClientTrace{
+		DNSStart: func(di httptrace.DNSStartInfo) {
+			dnsStart = time.Now()
+		},
+		DNSDone: func(di httptrace.DNSDoneInfo) {
+			report := dnsReport{
+				QueryName:  hostname,
+				Time:       dnsStart.UTC().Truncate(time.Second),
+				DurationMs: time.Since(dnsStart).Milliseconds(),
+			}
+			if di.Err != nil {
+				report.Error = di.Err.Error()
+			}
+			for _, ip := range di.Addrs {
+				report.AnswerIPs = append(report.AnswerIPs, ip.IP.String())
+			}
+			// TODO(fortuna): Use a Mutex.
+			r.DNSQueries = append(r.DNSQueries, report)
+		},
+		ConnectStart: func(network, addr string) {
+			connectStart = time.Now()
```

Comment:
This is not the correct behavior. In Happy Eyeballs we can have overlapping connection attempts.

We need a map network, addr -> start time instead.